### PR TITLE
Adding MDC themes

### PR DESCRIPTION
**Pull request by jgunaratne**
Proposal for MDC theme component. MDC themes make use of color schemes (design terminology). The theme class has been renamed color scheme to be consistent with design terminology. In the future the theme component directory could include other theme-able aspects of the UI. The mutator class continues to be named "themer." There appears to be a preference to avoid the generic term mutator and refer to mutators according to what they mutate -- in this case themes.

Continues work from "Proposed component theme additions #1362"

### CODE REVIEW

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.m [hunk 1]
```diff
@@ -29,12 +26,14 @@ @interface MDCFeatureHighlightViewController () <UIViewControllerTransitioningDe
...
   MDCFeatureHighlightAnimationController *_animationController;
   MDCFeatureHighlightCompletion _completion;
-  MDCFeatureHighlightView *_featureHighlightView;
   NSTimer *_pulseTimer;
   UIView *_displayedView;
   UIView *_highlightedView;
 }
 
+@synthesize outerHighlightColor = _outerHighlightColor;
+@synthesize innerHighlightColor = _innerHighlightColor;
```
Comment (samnm) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.m]: I don't think you should duplicate the color state in the view controller. Just make the getters return view.highlightColor.

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.m [hunk 2]
```diff
@@ -87,32 +88,35 @@ - (nonnull instancetype)initWithHighlightedView:(nonnull UIView *)highlightedVie
...
-
+  
   self.view = _featureHighlightView;
 }
 
+- (void)viewWillLayoutSubviews {
+  _featureHighlightView.titleLabel.text = self.titleText;
+  _featureHighlightView.bodyLabel.text = self.bodyText;
+  _featureHighlightView.outerHighlightColor = self.outerHighlightColor;
+  _featureHighlightView.innerHighlightColor = self.innerHighlightColor;
```
Comment (samnm) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.m]: If you deduplicate the color state you can remove these two color assignments.

File: components/Themes/src/MDCFeatureHighlightColorThemer.m [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//  featureHighlightController.innerHighlightColor = colorScheme.primaryColorLight;
+//  featureHighlightController.outerHighlightColor = colorScheme.primaryColor;
+//}
+
++ (void)applyColorScheme:(MDCColorScheme *)colorScheme
+    toFeatureHighlightView:(MDCFeatureHighlightView *)featureHighlightView {
+  NSLog(@"%@", featureHighlightView.outerHighlightColor);
+  featureHighlightView.innerHighlightColor = colorScheme.primaryColorLight;
+  featureHighlightView.outerHighlightColor = colorScheme.primaryColor;
+  NSLog(@"%@", featureHighlightView.outerHighlightColor);
```
Comment (samnm) [inline components/Themes/src/MDCFeatureHighlightColorThemer.m]: Remove logs.

File: components/Themes/examples/FeatureHighlightColorThemerTypicalUseViewController.m [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  MDCColorScheme *colorScheme = [[MDCColorScheme alloc] init];
+  colorScheme.primaryColor = [MDCPalette tealPalette].tint500;
+  colorScheme.primaryColorLight = [MDCPalette tealPalette].tint100;
+  colorScheme.primaryColorDark = [MDCPalette tealPalette].tint700;
+  colorScheme.secondaryColor = [MDCPalette bluePalette].tint500;
+  colorScheme.secondaryColorLight = [MDCPalette bluePalette].tint100;
+  colorScheme.secondaryColorDark = [MDCPalette bluePalette].tint700;
+
+  [MDCFeatureHighlightColorThemer applyColorScheme:colorScheme
+                            toFeatureHighlightView:[MDCFeatureHighlightView appearance]];
```
Comment (samnm) [inline components/Themes/examples/FeatureHighlightColorThemerTypicalUseViewController.m]: Ah this is cool. I didn't realize you could treat the appearance proxy as a native instance. Makes sense :)

**Review by samnm — DISMISSED** [review]

File: components/Themes/src/MDCColorScheme.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ limitations under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+@interface MDCColorScheme : NSObject
+
+// Material design guidelines recommend using primary and secondary colors with light and dark color
+// variants.
+@property (nonatomic, strong) UIColor *primaryColor;
```
Comment (ajsecord) [inline components/Themes/src/MDCColorScheme.h]: Docs for every public class, property, and method, please!

File: components/Themes/src/MDCFeatureHighlightColorThemer.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+#import "MaterialFeatureHighlight.h"
+#import "MDCColorScheme.h"
+
+@interface MDCFeatureHighlightColorThemer : NSObject
```
Comment (ajsecord) [inline components/Themes/src/MDCFeatureHighlightColorThemer.h]: docs docs docs docs docs 😉

**Review by ajsecord — DISMISSED** [review]

File: components/Themes/src/MDCColorScheme.h (hunk 5 above)
Comment (jgunaratne) [reply components/Themes/src/MDCColorScheme.h]: Done. Added documentation in comment form.

File: components/Themes/src/MDCFeatureHighlightColorThemer.h (hunk 6 above)
Comment (jgunaratne) [reply components/Themes/src/MDCFeatureHighlightColorThemer.h]: Done. Added docs.

**Review by ianegordon — APPROVED** [review]

**Comment by ianegordon** [comment]
Waiting for Travis, but LGTM.

File: components/Dialogs/examples/DialogsAlertViewController.m [hunk 7]
```diff
@@ -47,11 +46,7 @@ - (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPa
   }
 }
 
-
-
 - (IBAction)didTapShowAlert:(id)sender {
-  [[MDCButton appearanceWhenContainedIn:[MDCAlertController class], nil]
```
Comment (ianegordon) [inline components/Dialogs/examples/DialogsAlertViewController.m:53]: Typically this is done in appDidFinishLoading.
In this case, I would do it in viewDidLoad.

File: components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  UITapGestureRecognizer *tapRecognizer =
+      [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didTapBackground:)];
+  [self.view addGestureRecognizer:tapRecognizer];
+}
+
+- (void)viewDidLayoutSubviews {
+  [super viewDidLayoutSubviews];
+
+  [self.button sizeToFit];
```
Comment (ianegordon) [inline components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m]: nit: Redundant, already sized above.

File: components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  CGRect frame = self.button.frame;
+  frame.origin.x = self.view.frame.size.width / 2 - frame.size.width / 2;
```
Comment (ianegordon) [inline components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m]: frame may not be standardized (They may have negative values for width and height).
Use almost always want to use CGRectGetWidth(self.view.frame) when using frames.

File: components/Themes/src/MDCColorScheme.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+@interface MDCColorScheme : NSObject
+
+/** The main, primary color used for a theme. */
+@property (nonatomic, strong) UIColor *primaryColor;
+
+/** 
+ A slightly lighter version of the primary color. Given tonal variations of a color, this color is
+ typically two color swatches lighter than the primary color.
+ */
+@property (nonatomic, strong) UIColor *primaryColorLight;
```
Comment (ianegordon) [inline components/Themes/src/MDCColorScheme.h]: The type typically comes at the end, so this would be primaryLightColor.
Here and below.

**Review by ianegordon — DISMISSED** [review]

File: components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m (hunk 8 above)
Comment (jgunaratne) [reply components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m]: Done.

File: components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m (hunk 9 above)
Comment (jgunaratne) [reply components/Themes/examples/supplemental/FeatureHighlightColorThemerSupplemental.m]: Done.

File: components/Themes/src/MDCColorScheme.h (hunk 10 above)
Comment (jgunaratne) [reply components/Themes/src/MDCColorScheme.h]: Done.

File: components/Dialogs/examples/DialogsAlertViewController.m (hunk 7 above)
Comment (jgunaratne) [reply components/Dialogs/examples/DialogsAlertViewController.m:53]: Discussed offline. For the example, the theme will be applied immediately before the alert is presented.

**Review by ianegordon — APPROVED** [review]

File: MaterialComponents.podspec [hunk 11]
```diff
@@ -269,6 +269,15 @@ Pod::Spec.new do |s|
     ss.dependency "MaterialComponents/private/RTL"
   end
 
+  s.subspec "Themes" do |ss|
+    ss.ios.deployment_target = '8.0'
+    ss.public_header_files = "components/#{ss.base_name}/src/*.h"
+    ss.source_files = "components/#{ss.base_name}/src/*.{h,m}", "components/#{ss.base_name}/src/private/*.{h,m}"
+    ss.resources = ["components/#{ss.base_name}/src/Material#{ss.base_name}.bundle"]
+
+    ss.dependency "MaterialComponents/FeatureHighlight"
```
Comment (randallli) [inline MaterialComponents.podspec:278]: This makes the themes dependent on FeatureHighlight. This is not so scalable if the themes depend on all the components.

The best alternative I can come up with is a separate subspec for the themer.

**Pull request merged**